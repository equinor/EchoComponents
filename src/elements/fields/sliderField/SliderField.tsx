import { Slider, Typography } from '@equinor/eds-core-react';
import React, { CSSProperties } from 'react';
import styles from './sliderField.module.css';

export interface SliderFieldProps {
    value: number;
    min: number;
    max: number;
    title: string;
    onChange: (value: number) => void;
    style?: CSSProperties;
}

export const SliderField: React.FC<SliderFieldProps> = ({
    value,
    min,
    max,
    title,
    onChange,
    style
}: SliderFieldProps): JSX.Element => {
    return (
        <div className={styles.sliderField} style={style}>
            <Typography variant="body_short">{title}</Typography>
            <div className={styles.sliderWrapper}>
                <Slider
                    value={value}
                    min={min}
                    max={max}
                    ariaLabelledby={title}
                    onChange={(e: React.FormEvent<HTMLDivElement>, newValue: number | number[]): void => {
                        onChange(newValue as number);
                    }}
                />
            </div>
        </div>
    );
};
