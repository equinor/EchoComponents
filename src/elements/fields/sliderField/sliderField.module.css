.sliderField {
    margin: 0 0 30px 0;
    display: flex;
    flex-direction: column;
}

.sliderField p,
.sliderField span,
.sliderField output {
    font-weight: 500;
}

.sliderWrapper {
    width: 230px;
    text-align: center;
    margin: 10px auto 0 auto;
}

.sliderWrapper span {
    display: none;
}

.sliderWrapper output {
    font-size: 16px;
}
