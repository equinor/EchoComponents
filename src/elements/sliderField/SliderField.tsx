import { Slider, Typography } from '@equinor/eds-core-react';
import React, { CSSProperties } from 'react';
import styles from './sliderField.module.css';

export interface SliderFieldProps {
    value: number;
    min: number;
    max: number;
    labels?: string[];
    title: string;
    onChange: (value: number) => void;
    style?: CSSProperties;
}
/**
 * Component that renders a slider with numeric or custom values
 *
 * @param {SliderFieldProps} {
 *     value: Current value of the slider,
 *     min: Minimum value of the slider,
 *     max: Maximum value of the slider,
 *     labels: List of labels to use if the slider should not display numeric values,
 *     title: Text shown as title of the slider,
 *     onChange: Method that will be called if the slider value is changed,
 *     style: style element to override wrapper style
 * }
 * @return {*}  {JSX.Element}
 */
export const SliderField: React.FC<SliderFieldProps> = ({
    value,
    min,
    max,
    labels,
    title,
    onChange,
    style
}: SliderFieldProps): JSX.Element => {
    const outputFunctionForLabels = (currentValue: number, labelsAsParam: string[]): string => {
        return labelsAsParam[currentValue];
    };

    return (
        <div className={styles.sliderField} style={style}>
            <Typography variant="body_short">{title}</Typography>
            <div className={styles.sliderWrapper}>
                {!labels || (labels && labels.length === 0) ? (
                    <Slider
                        value={value}
                        min={min}
                        max={max}
                        ariaLabelledby={title}
                        onChange={(e: React.FormEvent<HTMLDivElement>, newValue: number | number[]): void => {
                            onChange(newValue as number);
                        }}
                    />
                ) : (
                    <Slider
                        value={value}
                        min={0}
                        max={labels.length - 1}
                        ariaLabelledby={title}
                        outputFunction={(currentNr: number): string => outputFunctionForLabels(currentNr, labels)}
                        onChange={(e: React.FormEvent<HTMLDivElement>, newValue: number | number[]): void => {
                            onChange(newValue as number);
                        }}
                    />
                )}
            </div>
        </div>
    );
};
