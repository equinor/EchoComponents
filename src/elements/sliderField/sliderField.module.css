.sliderField {
    margin: 0 0 30px 0;
    display: flex;
    flex-direction: column;
}

.sliderField p,
.sliderField span,
.sliderField output {
    font-weight: 500;
}

.sliderWrapper {
    width: 230px;
    text-align: center;
    margin: 10px auto 0 auto;
}

.sliderWrapper span {
    display: none;
}

.sliderWrapper output {
    font-size: 16px;
}

.sliderWrapper div {
    background-image: url("data:image/svg+xml,<svg xmlns='http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg'><rect x='0' y='11' fill='rgb(222, 237, 238)' width='100%' height='4' rx='2' /></svg>");
}
