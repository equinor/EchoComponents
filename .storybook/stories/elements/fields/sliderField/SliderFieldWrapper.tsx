import React from 'react';
import { SliderField, SliderFieldProps } from '../../../../../src/elements/fields/sliderField/SliderField';

const SliderFieldWrapper: React.FC<SliderFieldProps> = ({
    value,
    min,
    max,
    title,
    onChange,
    style
}: SliderFieldProps) => {
    return (
        <div className={'SliderField'}>
            <SliderField
                value={value}
                min={min}
                max={max}
                title={title}
                onChange={onChange}
                style={style}
            ></SliderField>
        </div>
    );
};

export default SliderFieldWrapper;
