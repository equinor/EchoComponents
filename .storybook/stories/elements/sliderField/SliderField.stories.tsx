import { Meta, Story } from '@storybook/react/types-6-0';
import React from 'react';
import { SliderFieldProps } from '../../../../src/elements/sliderField/SliderField';
import SliderFieldWrapper from './SliderFieldWrapper';

export default {
    title: 'Elements/Functional/SliderField',

    component: SliderFieldWrapper,
    argTypes: {}
} as Meta;

const Template: Story<SliderFieldProps> = (args) => <SliderFieldWrapper {...args} />;

export const Default = Template.bind({});

Default.args = {
    value: 15,
    min: 0,
    max: 20,
    title: 'Movement speed',
    onChange: () => {},
    style: { width: '350px' }
};
