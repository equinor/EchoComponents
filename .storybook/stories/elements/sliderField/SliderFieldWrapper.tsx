import React, { CSSProperties } from 'react';
import { SliderField } from '../../../../src/elements/sliderField/SliderField';

export enum OptionType {
    NUMERIC = 'Numeric',
    CUSTOM = 'Custom'
}
export interface SliderFieldWrapperProps {
    value: number;
    min: number;
    max: number;
    labels?: string[];
    title: string;
    onChange: (value: number) => void;
    style?: CSSProperties;
    variant: OptionType;
}

const SliderFieldWrapper: React.FC<SliderFieldWrapperProps> = ({
    value,
    min,
    max,
    labels,
    title,
    onChange,
    style,
    variant
}: SliderFieldWrapperProps) => {
    return (
        <div className={'SliderField'}>
            <SliderField
                value={value}
                min={min}
                max={max}
                labels={variant === OptionType.CUSTOM ? labels : undefined}
                title={variant === OptionType.CUSTOM ? 'Display Size' : title}
                onChange={onChange}
                style={style}
            ></SliderField>
        </div>
    );
};

export default SliderFieldWrapper;
