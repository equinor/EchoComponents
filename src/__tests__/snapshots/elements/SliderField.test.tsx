import React from 'react';
import renderer from 'react-test-renderer';
import { SliderField } from '../../../elements/sliderField/SliderField';

it('renders SliderField with numeric values correctly ', () => {
    const tree = renderer
        .create(<SliderField value={1} min={0} max={5} title={'New slider'} onChange={jest.fn()} />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders SliderField with labels correctly', () => {
    const tree = renderer
        .create(
            <SliderField
                value={2}
                min={0}
                max={3}
                title={'New slider with labels'}
                labels={['A', 'B', 'C', 'D']}
                onChange={jest.fn()}
            />
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
