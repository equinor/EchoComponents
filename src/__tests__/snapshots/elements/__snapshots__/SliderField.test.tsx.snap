// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SliderField with labels correctly 1`] = `
<div
  className="sliderField"
>
  <p
    className="sc-eCssSg bHjGEK"
  >
    New slider with labels
  </p>
  <div
    className="sliderWrapper"
  >
    <div
      className="sc-aemoO iIASIk"
      max={3}
      min={0}
      value={2}
    >
      <input
        aria-labelledby="New slider with labels"
        className="sc-TmcTc kXaNaA"
        id="New slider with labels-thumb"
        max={3}
        min={0}
        onChange={[Function]}
        onKeyUp={[Function]}
        onMouseUp={[Function]}
        step={1}
        type="range"
        value={2}
      />
      <output
        className="sc-dwfUOf cBYssx"
        htmlFor="New slider with labels-thumb"
        value={2}
      >
        C
      </output>
      <div
        className="sc-jONnTn sc-hlTvYk lpkCeB ciFSdt"
      />
      <span
        className="sc-fWSCIC cUIkTY"
      >
        A
      </span>
      <span
        className="sc-fWSCIC cUIkTY"
      >
        D
      </span>
    </div>
  </div>
</div>
`;

exports[`renders SliderField with numeric values correctly  1`] = `
<div
  className="sliderField"
>
  <p
    className="sc-eCssSg bHjGEK"
  >
    New slider
  </p>
  <div
    className="sliderWrapper"
  >
    <div
      className="sc-aemoO bKeWSN"
      max={5}
      min={0}
      value={1}
    >
      <input
        aria-labelledby="New slider"
        className="sc-TmcTc kXaNaA"
        id="New slider-thumb"
        max={5}
        min={0}
        onChange={[Function]}
        onKeyUp={[Function]}
        onMouseUp={[Function]}
        step={1}
        type="range"
        value={1}
      />
      <output
        className="sc-dwfUOf lfWHNa"
        htmlFor="New slider-thumb"
        value={1}
      >
        1
      </output>
      <div
        className="sc-jONnTn sc-hlTvYk lpkCeB ciFSdt"
      />
      <span
        className="sc-fWSCIC cUIkTY"
      >
        0
      </span>
      <span
        className="sc-fWSCIC cUIkTY"
      >
        5
      </span>
    </div>
  </div>
</div>
`;
